core/jvm/src/Buffer.kt

```diff
@@ -21,557 +21,281 @@
+    offset: Int,
+    byteCount: Int
+  ): Buffer = commonWrite(source, offset, byteCount)
 
-  override fun readUtf8() = readString(size, Charsets.UTF_8)
+  override fun writeAll(source: RawSource): Long = commonWriteAll(source)
 
-  @Throws(EOFException::class)
-  override fun readUtf8(byteCount: Long) = readString(byteCount, Charsets.UTF_8)
+  actual override fun write(source: RawSource, byteCount: Long): Buffer =
+    commonWrite(source, byteCount)
 
-  override fun readString(charset: Charset) = readString(size, charset)
+  actual override fun writeByte(byte: Byte): Buffer = commonWriteByte(byte)
 
-  @Throws(EOFException::class)
-  override fun readString(byteCount: Long, charset: Charset): String {
-    require(byteCount >= 0 && byteCount <= Integer.MAX_VALUE) { "byteCount: $byteCount" }
-    if (size < byteCount) throw EOFException()
-    if (byteCount == 0L) return ""
+  actual override fun writeShort(short: Short): Buffer = commonWriteShort(short)
 
-    val s = head!!
-    if (s.pos + byteCount > s.limit) {
-      // If the string spans multiple segments, delegate to readBytes().
-      return String(readByteArray(byteCount), charset)
-    }
+  actual override fun writeInt(int: Int): Buffer = commonWriteInt(int)
 
-    val result = String(s.data, s.pos, byteCount.toInt(), charset)
-    s.pos += byteCount.toInt()
-    size -= byteCount
+  actual override fun writeLong(long: Long): Buffer = commonWriteLong(long)
 
-    if (s.pos == s.limit) {
-      head = s.pop()
-      SegmentPool.recycle(s)
-    }
+  internal actual fun writableSegment(minimumCapacity: Int): Segment =
+    commonWritableSegment(minimumCapacity)
 
-    return result
-  }
+  override fun write(source: Buffer, byteCount: Long): Unit = commonWrite(source, byteCount)
 
-  @Throws(EOFException::class)
-  override fun readUtf8Line(): String? = commonReadUtf8Line()
+  override fun read(sink: Buffer, byteCount: Long): Long = commonRead(sink, byteCount)
 
-  @Throws(EOFException::class)
-  override fun readUtf8LineStrict() = readUtf8LineStrict(Long.MAX_VALUE)
+  public actual override fun flush(): Unit = Unit
 
-  @Throws(EOFException::class)
-  override fun readUtf8LineStrict(limit: Long): String = commonReadUtf8LineStrict(limit)
+  actual override fun close(): Unit = Unit
 
-  @Throws(EOFException::class)
-  override fun readUtf8CodePoint(): Int = commonReadUtf8CodePoint()
+  /**
+   * Returns a human-readable string that describes the contents of this buffer. Typically, this
+   * is a string like `[text=Hello]` or `[hex=0000ffff]`.
+   */
+  override fun toString(): String = commonString()
 
-  override fun readByteArray() = commonReadByteArray()
+  public actual fun copy(): Buffer = commonCopy()
 
-  @Throws(EOFException::class)
-  override fun readByteArray(byteCount: Long): ByteArray = commonReadByteArray(byteCount)
+  /**
+   * Returns a deep copy of this buffer.
+   *
+   * This method is equivalent to [copy].
+   */
+  public override fun clone(): Buffer = copy()
```

Comment:
Do we really want to pursue `Cloneable` here?

Reply:
@swankjesse Do you remember why it was implemented in the Okio in the first place (https://github.com/square/okio/commit/dad305d3c65647a9ea050f061a3544b0f4a67fed#diff-00883708f3cfa69f269509f3ade90c4e09e67bbd2a7a488e040cb82812284ba1R42)?

Reply:
Being able to deep-copy a Buffer is nice and I’d keep that behavior.

As for implementing Cloneable, I very rarely see APIs that do anything with `Cloneable` so I recommend dropping it.